rfcs/0053-pull-request-workflow.md

```diff
@@ -0,0 +1,93 @@
+---
+feature: pull-request workflow
+start-date: 2019-09-28
+author: Ingolf Wagner (@mrVanDalo)
+co-authors: lassulus (@lassulus)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: 
+---
+
+# Summary
+[summary]: #summary
+
+Pull-Request on GitHub are the main way we improve our code base.
+This Document should settle everything that needs to be done to bring code upstream.
+
+# Motivation
+[motivation]: #motivation
+
+Eliminate questions from contributors and maintainers
+about what should be done next and who should do it.
+This is not a new approach, it is more a settlement
+on how we do it now.
+This RFC is based on my experiences and talks I had
+with other contributors and maintainers.
+
+# Detailed design
+[design]: #detailed-design
+
+Define all steps of a Pull-Request.
+Use Roles to define responsibilities in every step.
+
+## Roles
+[roles]: #roles
+
+Everybody involved in the process of contributing has one or multiple
+of the following roles
+
+* Contributor
+* Bot
+* Reviewer
+* Maintainer (has merge privileges)
+
+The responsibilities of theses roles are defined in the rest of this RFC.
+
+## States of a Pull-Request
+[state]:#states
+
+This diagram defines all states of a Pull-Request,
+and their transitions to other states.
+
+![pull-request state](0053-pull-request-workflow/pull-request-states.svg)
+
+## Responsibilities and Actions
+[responsibilities]:#responsibilities
+
+The responsibilities of every role is defined by the following diagram:
+
+![pull-request activity](0053-pull-request-workflow/pull-request-roles.svg)
+
+## About Pull-Request
+
+### Packages
+
+* contributor must decide if a Backport is necessary
+* after the Pull-Request to master is merged the Backport Pull-Request is created
+* Backport Pull-Requests must be linked to the original Pull-Requests in master
+* reviewer and maintainer can deny the Backport
+
+### Modules
+
+* modules should have tests
+* new modules must have tests
```

Comment:
Tbh I'm not sure if that rule is too strict: I remember several cases where I actually wanted to write a test, but failed to find a suitable approach. One example that comes to my mind right now would be the [autorandr module](https://nixos.org/nixos/options.html#services.autorandr) where you'd have to simulate several monitors in the test VM.

I'd rather go with "reviewers should encourage contributors to write tests for new modules" to make it clear that we want to have tests but if there's no way around this, it's still fine to skip that task.

Reply:
True, it is not possible to create module tests for everything.

> reviewers should encourage contributors to write tests for new modules

Implies it is the responsibility of the reviewer to make sure tests are written (if possible). I can live with that.

Reply:
I guess another approach is to say that «new modules should have tests or an explicit description of problems with testing (e.g. hardware interaction)»